staging/src/k8s.io/client-go/testing/internal/listandwatch_test.go

```diff
@@ -0,0 +1,119 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+import (
+	"context"
+	"testing"
+	"testing/synctest"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2/ktesting"
+	_ "k8s.io/klog/v2/ktesting/init" // for -testing.v
+)
+
+// TestListAndWatch mirrors how fake client-go is often used with real
+// informers. It enforces a timing such that List completes, a new
+// object gets created because of the completed cache sync, and only
+// then is the Watch call in the reflector's "ListAndWatch" allowed to
+// continue.
+//
+// The fake Watch implementation then must use the ResourceVersion to
+// detect that it must send some (but not all!) objects to the new watch.
+//
+// This runs in a synctest bubble, therefore time is virtual.
+func TestListAndWatch(t *testing.T) { synctest.Test(t, testListAndWatch) }
+func testListAndWatch(t *testing.T) {
+	logger, ctx := ktesting.NewTestContext(t)
+	cm := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cm1",
+			Namespace: "default",
+		},
+	}
+	client := fake.NewSimpleClientset(cm)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	f := informers.NewSharedInformerFactory(client, 0)
+	configMapInformer := f.InformerFor(&v1.ConfigMap{}, func(client kubernetes.Interface, defaultEventHandlerResyncPeriod time.Duration) cache.SharedIndexInformer {
+
+		return cache.NewSharedIndexInformer(cache.ToListWatcherWithWatchListSemantics(&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				objs, err := client.CoreV1().ConfigMaps("").List(context.Background(), options)
+				logger.Info("Listed", "configMaps", objs, "err", err)
+				if err != nil {
+					t.Errorf("Unexpected List error: %v", err)
+				} else if objs.ResourceVersion != "1" {
+					t.Errorf("Expected ListMeta ResourceVersion 1, got %q", objs.ResourceVersion)
+				}
+				return objs, err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				if options.ResourceVersion != "1" {
+					t.Errorf("Expected ListOptions ResourceVersion 1, got %q", options.ResourceVersion)
+				}
+				logger.Info("Delaying Watch...")
+				// This blocks the goroutine without the need for
+				// extra synchronization and without slowing down
+				// the overall test duration.
+				time.Sleep(5 * time.Second)
```

Comment:
This cannot be done with a reactor because sleeping in the reactor also blocks the Create call.